x-pack/plugins/observability_solution/synthetics/server/routes/suggestions/route.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { SyntheticsRestApiRouteFactory } from '../types';
+import { syntheticsMonitorType } from '../../../common/types/saved_objects';
+import {
+  ConfigKey,
+  MonitorFiltersResult,
+  EncryptedSyntheticsMonitorAttributes,
+} from '../../../common/runtime_types';
+import { SYNTHETICS_API_URLS } from '../../../common/constants';
+import { QuerySchema, getMonitorFilters, SEARCH_FIELDS } from '../common';
+import { getAllLocations } from '../../synthetics_service/get_all_locations';
+
+type Buckets = Array<{
+  key: string;
+  doc_count: number;
+}>;
+
+interface AggsResponse {
+  locationsAggs: {
+    buckets: Buckets;
+  };
+  tagsAggs: {
+    buckets: Buckets;
+  };
+  projectsAggs: {
+    buckets: Buckets;
+  };
+}
+
+export const getSyntheticsSuggestionsRoute: SyntheticsRestApiRouteFactory<
+  MonitorFiltersResult
+> = () => ({
+  method: 'GET',
+  path: SYNTHETICS_API_URLS.SUGGESTIONS,
+  validate: {
+    query: QuerySchema,
+  },
+  handler: async (route): Promise<any> => {
+    const { savedObjectsClient } = route;
+    const { tags, locations, projects, monitorQueryIds, query } = route.request.query;
+
+    const { filtersStr } = await getMonitorFilters({
+      tags,
+      locations,
+      projects,
+      monitorQueryIds,
+      context: route,
+    });
+    const { allLocations = [] } = await getAllLocations(route);
+    const data = await savedObjectsClient.find<EncryptedSyntheticsMonitorAttributes>({
+      type: syntheticsMonitorType,
+      perPage: 1000,
+      fields: ['name', 'id'],
+      filter: filtersStr ? `${filtersStr}` : undefined,
+      aggs,
+      search: query ? `${query}*` : undefined,
+      searchFields: SEARCH_FIELDS,
+    });
+
+    const { tagsAggs, locationsAggs, projectsAggs } = (data?.aggregations as AggsResponse) ?? {};
+
+    return {
+      monitorIds: data.saved_objects.map(({ attributes }) => ({
+        label: attributes.name,
+        value: attributes.id,
+        count: 1,
+      })),
+      tags:
+        tagsAggs?.buckets?.map(({ key, doc_count: count }) => ({
+          label: key,
+          value: key,
+          count,
+        })) ?? [],
+      locations:
+        locationsAggs?.buckets?.map(({ key, doc_count: count }) => ({
+          label: allLocations.find((location) => location.id === key)?.label || key,
```

Comment:
You can avoid O(2^n) here if you just group the allocations by location.id and does like:
```suggestion
          label: allLocationsMap[key]?.label || key,
```